src/InodeCache.cpp

```diff
@@ -0,0 +1,485 @@
+// top level is a hash table consisting of buckets. Each bucket contains entries
+// that are sorted in LRU order. Entries map from keys representing files to
+// cached hash results.
+//
+// Concurrent access is guarded by a mutex in each bucket.
+//
+// Current cache size is fixed and the given constants are considered large
+// enough for most projects. The size could be made configurable if there is a
+// demand for it.
+const uint32_t k_version = 1;
+
+// Increment version number if constants affecting storage size are changed.
+const uint32_t k_num_buckets = 32 * 1024;
+const uint32_t k_num_entries = 4;
+
+static_assert(sizeof(digest::bytes) == 20,
+              "Increment version number if size of digest is changed.");
+
+static_assert(
+  static_cast<int>(InodeCache::ContentType::binary) == 0,
+  "Numeric value is part of key, increment version number if changed.");
+static_assert(
+  static_cast<int>(InodeCache::ContentType::code) == 1,
+  "Numeric value is part of key, increment version number if changed.");
+static_assert(
+  static_cast<int>(InodeCache::ContentType::code_with_sloppy_time_macros) == 2,
+  "Numeric value is part of key, increment version number if changed.");
+
+std::string
+get_file_from_config(const Config& config)
+{
+  return (config.inode_cache_file().empty()
+            ? config.cache_dir() + "/inode_cache"
+            : config.inode_cache_file())
+         + ".v" + std::to_string(k_version);
+}
+
+} // namespace
+
+struct InodeCache::Key
+{
+  ContentType type;
+  dev_t st_dev;
+  ino_t st_ino;
+  mode_t st_mode;
+#  ifdef HAVE_STRUCT_STAT_ST_MTIM
+  timespec st_mtim;
+#  else
+  time_t st_mtim;
+#  endif
+#  ifdef HAVE_STRUCT_STAT_ST_CTIM
+  timespec st_ctim; // Included for sanity checking.
+#  else
+  time_t st_ctim; // Included for sanity checking.
+#  endif
+  off_t st_size; // Included for sanity checking.
+  bool sloppy_time_macros;
+};
+
+struct InodeCache::Entry
+{
+  digest key_digest;  // Hashed key
+  digest file_digest; // Cached file hash
+  int return_value;   // Cached return value
+};
+
+struct InodeCache::Bucket
+{
+  pthread_mutex_t mt;
+  Entry entries[k_num_entries];
+};
+
+struct InodeCache::SharedRegion
+{
+  uint32_t version;
+  std::atomic<int64_t> hits;
+  std::atomic<int64_t> misses;
+  std::atomic<int64_t> errors;
+  Bucket buckets[k_num_buckets];
+};
+
+bool
+InodeCache::mmap_file(const std::string& inode_cache_file)
+{
+  if (m_sr) {
+    munmap(m_sr, sizeof(SharedRegion));
+    m_sr = nullptr;
+  }
+  int fd = open(inode_cache_file.c_str(), O_RDWR);
+  if (fd < 0) {
+    cc_log("Failed to open inode cache %s: %s",
+           inode_cache_file.c_str(),
+           strerror(errno));
+    return false;
+  }
+  bool is_nfs;
+  if (Util::is_nfs_fd(fd, &is_nfs) == 0 && is_nfs) {
+    throw Error(fmt::format(
+      "Inode cache file \"{}\" must be located on a local filesystem.",
+      inode_cache_file));
+  }
```

Comment:
Without the check tests will not fail on NFS since they run on the same client. But the check is needed to avoid data corruption if running on multiple clients, since the mutexes will not work as expected.

I will change to /tmp for the tests to allow them to be run on nfs again.

Not sure about silently disabling the cache though. When I first tried ccache with ccache_dir on NFS I got terrible performance and it took a while before I figured out that temporary_dir should be local. As a user I would have preferred a failure with a hint on what to change instead. But I will do as you suggest if you are of a different opinion.

Reply:
> Without the check tests will not fail on NFS since they run on the same client.

Oh, this is getting confusing. Let's see.

I made my previous comment (some 4 weeks ago) on ec5b8adb048b4d405e7d437324453638c455b5fe about failing tests on NFS after actually testing on NFS. It looked like this:

```console
% unittest/run
[...]
unittest/test_InodeCache.cpp:81: FAILED:
  CHECK( InodeCache::zero_stats(config) )
with expansion:
  false

Failed to create new inode cache in directory /home/joel/.ccache: Operation not supported
[...]
% test/run inode_cache
[...]
Running test suite inode_cache.Failed to create new inode cache in directory /home/joel/tmp/ccache/testdir.2128382/.ccache: Operation not supported
```

I didn't investigate further and apparently subconsciously thought that `mmap` was the problem, but it wasn't. It was `O_TMPFILE`.

Now, on 4a0398023b3ff4b9be3587cb27f7ddeec9966c12, it looks like this with the NFS check disabled:

```console
% unittest/run
[...]
unittest/test_Util.cpp:158: FAILED:
  CHECK( Util::fallocate(fd, 5000) == 0 )
with expansion:
  9 == 0

unittest/test_Util.cpp:162: FAILED:
  CHECK( Util::fallocate(fd, 20000) == 0 )
with expansion:
  9 == 0

unittest/test_Util.cpp:164: FAILED:
  CHECK( Stat::stat(filename).size() == 20000 )
with expansion:
  10000 (0x2710) == 20000 (0x4e20)
% test/run inode_cache
[...]
Running test suite inode_cache.......
PASSED
```

This is because `posix_fallocate` (the `fallocate` syscall) isn't supported. However, glibc apparently emulates it for the _create file from scratch_ case but not for the _file already exists_ case. Perhaps we can just remove those `Util::fallocate` test cases? Or don't use `posix_fallocate`?

> Not sure about silently disabling the cache though. When I first tried ccache with ccache_dir on NFS I got terrible performance and it took a while before I figured out that temporary_dir should be local. As a user I would have preferred a failure with a hint on what to change instead. But I will do as you suggest if you are of a different opinion.

I would say that anything that is not a far-fetched edge case and that could fail people's builds with a default ccache installation is a no-no. One reason being the Fedora package's activate-by-default.

Reply:
Perhaps this makes putting the inode cache file in a local temp dir more attractive again?

Reply:
> This is because posix_fallocate (the fallocate syscall) isn't supported. However, glibc apparently emulates it for the create file from scratch case but not for the file already exists case. Perhaps we can just remove those Util::fallocate test cases? Or don't use posix_fallocate?

I could reproduce the failure in ubuntu 16, but not in Fedora 32 or ubuntu 20. Perhaps a bug in glibc that is fixed in more recent versions? In that case disabling the test when posix_fallocate is available makes sense, since we are not testing glibc?

> Perhaps this makes putting the inode cache file in a local temp dir more attractive again?


What do you think about the following?
1. Change default location for temporary_dir to /run/user/$(id -u)/ccache-tmp if /run/user exists, otherwise keep it in ccache_dir/tmp.
2. Always put the inode cache file in temporary_dir and remove the configuration option for relocating it.
3. Disable the inode cache feature if temporary_dir is located on nfs.

That should provide decent nfs performance with default settings for all systems that have /run/user and we don't have to check ownership since /run/user should have restricted access already. And disabling inode cache in case (3) will not make things much worse since performance should already be bad by other reasons.

Reply:
> I could reproduce the failure in ubuntu 16, but not in Fedora 32 or ubuntu 20. Perhaps a bug in glibc that is fixed in more recent versions?

I tested on Fedora 31 with glibc-2.30-11.fc31.x86_64. Note that it only happens on NFS.

> In that case disabling the test when posix_fallocate is available makes sense, since we are not testing glibc?

Perhaps.

> What do you think about the following?
> 
> 1. Change default location for temporary_dir to /run/user/$(id -u)/ccache-tmp if /run/user exists, otherwise keep it in ccache_dir/tmp.
> 2. Always put the inode cache file in temporary_dir and remove the configuration option for relocating it.
> 3. Disable the inode cache feature if temporary_dir is located on nfs.

Sounds good, especially after 62e5c43270df (see below)! :+1:

Some background in case anybody is interested: It used to be the case that `CCACHE_TEMPDIR` had to be on the same partition as `CCACHE_DIR`, but I fixed that in ccache 3.0 and made `CCACHE_TEMPDIR` default to `$CCACHE_DIR/tmp` instead of `$CCACHE_DIR`. It also used to be the case that ccache easily leaked temporary files when e.g. pressing Ctrl+C to interrupt a build, which is why there is special "clean up tmp dir if it's in $CCACHE_DIR/tmp" code that runs once in a while. But this should no longer be a problem after last week's 62e5c43270df since files no longer should leak except if terminated with SIGKILL or so.

Reply:
> I tested on Fedora 31 with glibc-2.30-11.fc31.x86_64. Note that it only happens on NFS.

Same thing on Fedora 32 with glibc-2.31-2.fc32.x86_64, by the way.

Reply:
> > I tested on Fedora 31 with glibc-2.30-11.fc31.x86_64. Note that it only happens on NFS.
> 
> Same thing on Fedora 32 with glibc-2.31-2.fc32.x86_64, by the way.

Looks quite random then since it works for me both in Fedora 31 and Fedora 32, but not in Ubuntu 16.04. This with all three systems as virtual machines on the same host and against the same nfs server and file system. Will have a look at it in the debugger when I get some spare time.

Reply:
Of course. The nfs client can't emulate fallocate without read permission to the file. :-)

Reply:
Fixed now with proper file permissions in the fallocate test, changed default location of temporary_dir and the persistent file always in temporary_dir, or disabled in case of nfs.